### DISCUSSION TITLE
Moving precipitates at grain boundaries

### DISCUSSION THREAD

**Discussion by mdh23rct** [idaholab/moose#25563] [Q&A Modules: Phase field]
Dear MOOSE experts,
Hope you are doing great! Currently, I am trying to simulate moving spherical precipitates in grain boundaries. For that I have to include:
1. Multiple precipitates moving along the grain boundaries.
2. precipitate-precipitate interaction and precipitate-grain boundary interaction.

I am planning to use the phase-field module for that. Would you be kind to advice me any particular tutorials I could follow.
Also do I need to combine any other module for that!

I would appreciate your advice.

Thanks.

**Comment by GiudGiud** [comment]
@laagesen

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
I am currently trying to simulate random formation of precipitates in a polycrystal using phase field. However, when I run the random formation code, I am seeing some boundary problems. The nucleation is happening at the outer boundaries (not the grain boundaries) of the computational domain, specially at the junction between grains. This seems to be some sort of computational error. For your convenience, I have attached the image of boundary problem and mesh in this mail.
![boundary problem](https://github.com/idaholab/moose/assets/141943199/dbdb656a-dc92-45ad-b85f-eb7bb569e0a3)
![Mesh](https://github.com/idaholab/moose/assets/141943199/787bfcc4-3682-44b2-9c5b-8dfab9b0a656)

I am also attaching the random nucleation code I am using. 
```
[Modules]
    [./PhaseField]
      [./Conserved]
        [./c]
          free_energy = F
          mobility = M
          kappa = kappa_c
          solve_type = REVERSE_SPLIT
        [../]
      [../]
    [../]
  []
  
  [ICs]
    [./c_IC]
      type = ConstantIC
      variable = c
      value = 0.2
    [../]
  []
  
  [Materials]
    [./pfmobility]
      type = GenericConstantMaterial
      prop_names  = 'M kappa_c'
      prop_values = '1 25'
    [../]
    [./chemical_free_energy]
      # simple double well free energy
      type = DerivativeParsedMaterial
      property_name = Fc
      coupled_variables = 'c'
      constant_names       = 'barr_height  cv_eq'
      constant_expressions = '0.1          0'
      expression = 16*barr_height*c^2*(1-c)^2 #+0.01*(c*plog(c,0.005)+(1-c)*plog(1-c,0.005))
      derivative_order = 2
      outputs = exodus
    [../]
    [./probability]
      # This is a made up toy nucleation rate it should be replaced by
      # classical nucleation theory in a real simulation.
      type = ParsedMaterial
      property_name = P
      coupled_variables = c
      expression = 'if(c<0.21,c*1e-8,0)'
      outputs = exodus
    [../]
    [./nucleation]
      # The nucleation material is configured to insert nuclei into the free energy
      # tht force the concentration to go to 0.95, and holds this enforcement for 500
      # time units.
      type = DiscreteNucleation
      property_name = Fn
      op_names  = c
      op_values = 0.90
      penalty = 5
      penalty_mode = MIN
      map = map
      outputs = exodus
    [../]
    [./free_energy]
      # add the chemical and nucleation free energy contributions together
      type = DerivativeSumMaterial
      derivative_order = 2
      coupled_variables = c
      sum_materials = 'Fc Fn'
    [../]
  []
  
  [UserObjects]
    [./inserter]
      # The inserter runs at the end of each time step to add nucleation events
      # that happened during the timestep (if it converged) to the list of nuclei
      type = DiscreteNucleationInserter
      hold_time = 50
      probability = P
      radius = 1
    [../]
    [./map]
      # The map UO runs at the beginning of a timestep and generates a per-element/qp
      # map of nucleus locations. The map is only regenerated if the mesh changed or
      # the list of nuclei was modified.
      # The map converts the nucleation points into finite area objects with a given radius.
      type = DiscreteNucleationMap
      periodic = c
      inserter = inserter
    [../]
  []
  
  [Preconditioning]
    [./SMP]
      type = SMP
      full = true
    [../]
  []
  
  [BCs]
    [./Periodic]
      [./all]
        auto_direction = 'x y'
      [../]
    [../]
  []
  
  [Postprocessors]
    [./dt]
      type = TimestepSize
    [../]
    [./ndof]
      type = NumDOFs
    [../]
    [./rate]
      type = DiscreteNucleationData
      value = RATE
      inserter = inserter
    [../]
    [./dtnuc]
      type = DiscreteNucleationTimeStep
      inserter = inserter
      p2nucleus = 0.0005
      dt_max = 10
    [../]
    [./update]
      type = DiscreteNucleationData
      value = UPDATE
      inserter = inserter
    [../]
    [./count]
      type = DiscreteNucleationData
      value = COUNT
      inserter = inserter
    [../]
  []
  
  [Adaptivity]
    [./Indicators]
      [./jump]
        type = GradientJumpIndicator
        variable = c
      [../]
    [../]
    [./Markers]
      [./nuc]
        type = DiscreteNucleationMarker
        map = map
      [../]
      [./grad]
        type = ValueThresholdMarker
        variable = jump
        coarsen = 0.1
        refine = 0.2
      [../]
      [./combo]
        type = ComboMarker
        markers = 'nuc grad'
      [../]
    [../]
    marker = combo
    cycles_per_step = 3
    recompute_markers_during_cycles = true
    max_h_level = 3
  []
  
  [Executioner]
    type = Transient
    scheme = bdf2
    solve_type = 'PJFNK'
    petsc_options_iname = '-pc_type -sub_pc_type'
    petsc_options_value = 'asm      lu          '
  
    nl_max_its = 100
    l_tol = 1.0e-4
    nl_rel_tol = 1.0e-9
    nl_abs_tol = 1.0e-9
    start_time = 0.0
    end_time = 4000
  
    [./TimeStepper]
      type = IterationAdaptiveDT
      dt = 10
      growth_factor = 1.5
      cutback_factor = 0.5
      optimal_iterations = 8
      iteration_window = 2
      timestep_limiting_postprocessor = dtnuc
    [../]
  []
  
  [Outputs]
    exodus = true
    csv = true
    print_linear_residuals = false
  []
```
Would you be kind to let me know if I am missing something in the code to remove this boundary issues? Or the definition of the BCs is wrong in these code? I would appreciate your advice.

Best regards.

**Reply by GiudGiud** [reply]
This looks like a poor interaction between adaptivity and periodic boundary conditions.
Can you try to run the code in debug module (recompile with `METHOD=debug make -j4`) to see if you get an error?

Else would you consider turning off adaptivity near the borders? You d have to make a custom indicator probably

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Thank you for your reply. Since I am a newbie in MOOSE, please don't be annoyed by my questions:
1. How I can turn off adaptivity near the borders? As far as I guess, I have to generate a specific block (excluding the border regions), for which the adaptivity will be applied. Maybe you know better ways to do it? Please be aware that the mesh is irregular and not symmetric. Therefore, I exactly don't know how I can generate this mesh block!! 
2. How to make a custom indicator for the border region? The same way as it is done for the whole domain??

If you have any tutorials regarding these, I would be happy if you can forward.

I would appreciate your reply. Thanks.

**Reply by GiudGiud** [reply]
>  the mesh is irregular and not symmetric

it s periodic though right? If it s not periodic, the periodic BCs will fail

**Reply by mdh23rct** [reply]
Yes, the mesh is periodic. However, mesh adaptivity has some issue with the periodic boundary condition. If I turn off the periodic boundary condition, the border problem disappears.

**Reply by GiudGiud** [reply]
@roystgnr do we support adaptivity with a periodic mesh that becomes non-periodic with adaptivity?

**Reply by roystgnr** [reply]
We definitely should support consistently combined constraints when using adaptivity with a periodic mesh.  `git grep -l auto_direction | xargs grep -l Adaptivity` shows a few dozen test files (mostly in the phase field module) using both.

The results here are obviously a non-physical bug, and almost certainly due to some adaptivity+periodicity interaction, but I don't understand what "extra" is going on that we're not covering in regression tests.

**Reply by laagesen** [reply]
Hi @mdh23rct , if you want to insert nuclei at grain boundaries, you are going to need to modify this example- right now you are setting c = 0.2 everywhere in the initial conditions, and then making the nucleation probability uniformly high everywhere c < 0.21. So, there is nothing that makes the nucleation localized to GBs as far as I can tell. I am not sure why it seems to be happening at the domain boundaries first, but the problem is not set up to capture the physics you want so I would start by trying to correct that. How are the grain boundaries represented in your simulation? It seems that you are running on a mesh that is refined near GBs, but that is not enough to represent the grain boundaries in the simulation. Normally we do this by adding some additional order parameters to represent the presence of the grains.

**Reply by mdh23rct** [reply]
Dear Dr. Larry,
I appreciate your reply. 
1. We did SEM on the specimens. We are seeing carbide formation predominantly at the grain boundaries. That means carbon is diffusing to the grain boundaries.
2. Ok, I will modify the file as you adviced, and keep you updated.

Thanks for your time and support. I appreciate.
